quizapp/src/main/java/billennium/quizapp/service/QuizService.java

```diff
@@ -2,38 +2,97 @@
 import billennium.quizapp.entity.QuizDefinition;
+import billennium.quizapp.entity.QuizExecuted;
+import billennium.quizapp.entity.QuizStatus;
+import billennium.quizapp.entity.ResultDetails;
+import billennium.quizapp.exception.QuizDefinitionException;
+import billennium.quizapp.exception.QuizExecutedException;
 import billennium.quizapp.projection.QuizExecutedView;
 import billennium.quizapp.repository.CandidateRepository;
+import billennium.quizapp.repository.QuizExecutedRepository;
 import billennium.quizapp.resource.quiz.AnswerDto;
+import billennium.quizapp.resource.quiz.AnswersDto;
 import billennium.quizapp.resource.quiz.QuestionDto;
 import billennium.quizapp.resource.quiz.QuizDefinitionDto;
 import lombok.RequiredArgsConstructor;
 import org.modelmapper.ModelMapper;
 import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
 
 import java.util.List;
 import java.util.UUID;
 import java.util.stream.Collectors;
 
 @Service
 @RequiredArgsConstructor
+@Transactional
 public class QuizService {
-    private final CandidateRepository candidateRepository;
 
-    private final static Integer QUIZ_FIRST_QUESTION_NUMBER = 0;
+    private final CandidateRepository candidateRepository;
+    private final QuizExecutedRepository quizExecutedRepository;
 
-    public QuizDefinitionDto getFirstQuizQuestion(String userId) {
+    public QuizDefinitionDto begin(String userId, AnswersDto answersDto) {
         QuizExecutedView queryResult = candidateRepository.getCandidateWithQuiz(UUID.fromString(userId));
         if (queryResult != null) {
-            return mapToQuizDefinitionModel(queryResult.getQuizExecutedId(), queryResult.getQuiz());
+            return mapToQuizDefinitionModel(queryResult.getQuizExecutedId(), queryResult.getQuiz(), answersDto);
         } else {
             return new QuizDefinitionDto();
         }
     }
 
+    private QuizDefinitionDto mapToQuizDefinitionModel(Long quizExecutedId, QuizDefinition quiz,
+                                                       AnswersDto answersDTO) {
+        if (answersDTO.getQuestionId() == null) {
+            createQuizExecuted(quiz);
+            Question firstQuestion = getFirstQuestion(quiz);
+            return createQuizDefinitionDto(quizExecutedId, firstQuestion, quiz, firstQuestion.getId());
+        } else {
+            return goToNextQuestion(quizExecutedId, answersDTO, quiz);
+        }
+    }
+
+
+    private QuizDefinitionDto goToNextQuestion(Long quizExecutedId, AnswersDto answersDTO, QuizDefinition quiz) {
+        QuizExecuted quizExecuted = quizExecutedRepository.findById(quizExecutedId).orElseThrow(QuizExecutedException::new);
+        List<ResultDetails> resultDetails = saveGivenAnswer(answersDTO, quizExecuted);
+        return createQuizDefinitionDtoForNextQuestion(resultDetails, quiz, quizExecuted);
+    }
+
+    private List<ResultDetails> saveGivenAnswer(AnswersDto answersDTO, QuizExecuted quizExecuted) {
```

Comment:
metoda nazywa się `saveGivenAnswer` a nie widze w kodzie metody nic związanego z zapisywaniem. metoda powinna nazywać się wg mnie `buildResultDetails` lub coś podobnego. To co zwraca metoda jest zueplnie inne niż Twoja nazwa.

Reply:
i oczywiście nie trzeba przypisywać listy do obiektu, dodać coś do niej i potem ją zwrócic tylko można przecież od razu zwrócić. Niepotrzebna alokacja pamięci

Reply:
w tej metodzie trochę  się dzieje :) bo jest zapis udzielonych odpowiedzi do tabelki ResultDetails. 
Stąd też jest w parametrze quizExecuted.
Nie miałem póki co lepszego pomysłu na rozwiązanie tego.